src/java.base/share/classes/java/lang/StringCoding.java

```diff
@@ -97,29 +116,83 @@ public static int countPositives(byte[] ba, int off, int len) {
+     * @param sa the source byte array containing characters encoded in UTF-16
+     * @param sp the index of the <em>byte (not character!)</em> from the source array to start reading from
+     * @param da the target byte array
+     * @param dp the index of the target array to start writing to
+     * @param len the total number of <em>characters (not bytes!)</em> to be encoded
+     * @return the total number of <em>characters (not bytes!)</em> successfully encoded
+     * @throws NullPointerException if any of the provided arrays is null
+     * @throws ArrayIndexOutOfBoundsException if any of the provided sub-ranges are
+     *         {@linkplain Preconditions#checkFromIndexSize(int, int, int, BiFunction) out of bounds}
+     */
+    static int encodeISOArray(byte[] sa, int sp, byte[] da, int dp, int len) {
+        checkFromIndexSize(sp, len << 1, requireNonNull(sa, "sa").length, AIOOBE_FORMATTER);
+        checkFromIndexSize(dp, len, requireNonNull(da, "da").length, AIOOBE_FORMATTER);
+        return encodeISOArray0(sa, sp, da, dp, len);
+    }
+
     @IntrinsicCandidate
-    public static int implEncodeISOArray(byte[] sa, int sp,
-                                         byte[] da, int dp, int len) {
+    private static int encodeISOArray0(byte[] sa, int sp, byte[] da, int dp, int len) {
         int i = 0;
         for (; i < len; i++) {
             char c = StringUTF16.getChar(sa, sp++);
             if (c > '\u00FF')
                 break;
-            da[dp++] = (byte)c;
+            da[dp++] = (byte) c;
         }
         return i;
     }
 
+    /**
+     * Encodes as many ASCII codepoints as possible from the source
+     * character array into the destination byte array, assuming that
+     * the encoding is ASCII compatible.
+     *
+     * @param sa the source character array
+     * @param sp the index of the source array to start reading from
+     * @param da the target byte array
+     * @param dp the index of the target array to start writing to
+     * @param len the total number of characters to be encoded
+     * @return the total number of characters successfully encoded
+     * @throws NullPointerException if any of the provided arrays is null
+     * @throws ArrayIndexOutOfBoundsException if any of the provided sub-ranges are
+     *         {@linkplain Preconditions#checkFromIndexSize(int, int, int, BiFunction) out of bounds}
+     */
+    static int encodeAsciiArray(char[] sa, int sp, byte[] da, int dp, int len) {
+        checkFromIndexSize(sp, len, requireNonNull(sa, "sa").length, AIOOBE_FORMATTER);
+        checkFromIndexSize(dp, len, requireNonNull(da, "da").length, AIOOBE_FORMATTER);
+        return encodeAsciiArray0(sa, sp, da, dp, len);
+    }
+
     @IntrinsicCandidate
-    public static int implEncodeAsciiArray(char[] sa, int sp,
-                                           byte[] da, int dp, int len)
-    {
+    private static int encodeAsciiArray0(char[] sa, int sp, byte[] da, int dp, int len) {
```

Comment:
I rather prefer the readability of the src and dest parameters being on separate lines. (but not "{" by itself
```suggestion
    private static int encodeAsciiArray0(char[] sa, int sp, 
                                         byte[] da, int dp, int len) {
```